clang/test/SemaSYCL/builtin-fields.cpp

```diff
@@ -0,0 +1,164 @@
+// RUN: %clang_cc1 -fsycl-is-device -fsyntax-only -verify %s
+
+template<class T, T v>
+struct integral_constant {
+    static constexpr T value = v;
+    using value_type = T;
+    using type = integral_constant;
+    constexpr operator value_type() const noexcept { return value; }
+    constexpr value_type operator()() const noexcept { return value; }
+};
+
+using true_type = integral_constant<bool, true>;
+using false_type = integral_constant<bool, false>;
+
+template<class T, class U>
+struct is_same : false_type {};
+
+template<class T>
+struct is_same<T, T> : true_type {};
+
+struct A { int i; };
+struct B { float f; int i; A a; };
+struct C { struct { int a, b; }; };
+struct D { union { int a; float f; }; };
+union E { int i; float f; double d; };
+template <typename Ty> struct F { Ty a; int b; }; // expected-error {{field has incomplete type 'S'}}
+template <int N> struct G { int i = N; };
+
+// This class has two fields and one base. We don't treat base classes
+// as a field so that the SYCL library can generate more sensible
+// diagnostics.
+struct H : B {
+  double d;
+  const char *str;
+};
+
+struct I {};
+struct J { int &derp; const int &herp; int &&berp; };
+struct K { int a[10]; int b[]; };
+struct L { int a; int b : 1; int : 0; int c; };
+
+void easy() {
+  static_assert(__builtin_num_fields(A) == 1, "expected one field");
+  static_assert(__builtin_num_fields(B) == 3, "expected three fields");
+  static_assert(__builtin_num_fields(E) == 3, "expected three fields");
+  static_assert(__builtin_num_fields(H) == 2, "expected two fields");
+  static_assert(__builtin_num_fields(I) == 0, "expected no fields");
+  static_assert(__builtin_num_fields(J) == 3, "expected three fields");
+
+  static_assert(is_same<decltype(__builtin_field_type(A, 0)), int>::value, "expected an int");
+
+  static_assert(is_same<decltype(__builtin_field_type(B, 0)), float>::value, "expected a float");
+  static_assert(is_same<decltype(__builtin_field_type(B, 1)), int>::value, "expected an int");
+  static_assert(is_same<decltype(__builtin_field_type(B, 2)), A>::value, "expected an A");

```

Comment:
C presumably isn't assertable because it is the anonymous struct, right?

Reply:
Correct. The only thing I could think of would be to give the anonymous struct a declarator that we could use `decltype` on. e.g.,
```
struct C { struct { int a, b; } anon; };
 static_assert(is_same<decltype(__builtin_field_type(C, 0)), decltype(C{}.anon)>::value, "expected a something, but not certain what");
```
But I don't think that's quite the same test any longer because `a` and `b` in an anonymous struct would behave as if they're members of `C` but won't if we give the anonymous struct a declarator.

Reply:
You could maybe use nested calls to __builtin_field_type to see if the inner-types are still there.  so in that example:

````
struct C { struct { int a, b; } ; };
static_assert(is_same<decltype(__builtin_field_type(decltype(__builtin_field_type(C, 0)), 0), int>::value, "Isn't an int?");
````
I might have gotten lost in parens there :)

Reply:
Good idea! I'll add the tests.